WireAPI/Sources/WireAPI/Network/NetworkService/NetworkService.swift

```diff
@@ -0,0 +1,148 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+final class NetworkService: NSObject {
+
+    private let baseURL: URL
+    private var urlSession: URLSession?
+    private var webSocketsByTask = [URLSessionWebSocketTask: WebSocket]()
+
+    init(baseURL: URL) {
+        self.baseURL = baseURL
+    }
+
+    deinit {
+        urlSession?.invalidateAndCancel()
+    }
+
+    func configure(with urlSession: URLSession) {
+        self.urlSession = urlSession
+    }
+
+    func executeRequest(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
+        guard let urlSession else {
+            throw NetworkServiceError.serviceNotConfigured
+        }
+
+        guard let url = request.url else {
+            throw NetworkServiceError.invalidRequest
+        }
+
+        var request = request
+        request.url = URL(
+            string: url.absoluteString,
+            relativeTo: baseURL
+        )
+
+        let (data, response) = try await urlSession.data(for: request)
+
+        guard let httpURLResponse = response as? HTTPURLResponse else {
+            throw NetworkServiceError.notAHTTPURLResponse
+        }
```

Comment:
I don't know if you planned to handle it somewhere else but what do you think of handling errors more explicitly at the NetworkService level ?  Just an example here.

```
func mapResponse(response: (data: Data, response: URLResponse)) throws -> Data {
    guard let httpResponse = response.response as? HTTPURLResponse else {
        throw NetworkServiceError.notAHTTPURLResponse
    }
    
    switch httpResponse.statusCode {
    case 200..<300:
        return response.data
    case 400:
        throw NetworkServiceError.badRequest
    case 401:
        throw NetworkServiceError.unauthorized
    case 403:
        throw NetworkServiceError.forbidden
    case 404:
        throw NetworkServiceError.notFound
    default:
        throw NetworkServiceError.http(httpResponse: httpResponse, data: response.data)
    }
}
```

Reply:
Yes, for general errors that apply to any endpoint (such as authentication or rate limiting errors) then we can handle them here, but we have another ticket for that.